src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs

```diff
@@ -152,7 +152,7 @@ private void ComputeAndCacheId()
         /// </returns>
         private Uri ComputeId()
         {
-            if (this.ResourceMetadataContext.KeyProperties.Any())
+            if (this.ResourceMetadataContext.KeyProperties.Any() && this.ResourceMetadataContext.TypeContext.NavigationSourceName != null)

```

Comment:
move the  'null' testing ahead of 'Any()' could be better perf?

Reply:
:thumbsup: 

I like it!